addons/binding/org.openhab.binding.feican/src/main/java/org/openhab/binding/feican/internal/FeicanDiscoveryService.java

```diff
@@ -0,0 +1,196 @@
+
+import static org.openhab.binding.feican.FeicanBindingConstants.*;
+
+import java.io.IOException;
+import java.net.DatagramPacket;
+import java.net.DatagramSocket;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.net.SocketException;
+import java.net.SocketTimeoutException;
+import java.net.UnknownHostException;
+import java.nio.charset.StandardCharsets;
+import java.util.Locale;
+import java.util.Map;
+import java.util.TreeMap;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.config.discovery.DiscoveryService;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Discovery service for Feican Bulbs. When sending a discovery UDP broadcast command on port 5000 to a Feican bulb. The
+ * bulp will respond with it's mac address send via UDP broadcast over port 6000.
+ *
+ * @author Hilbrand Bouwkamp - Initial contribution
+ */
+@NonNullByDefault
+@Component(service = DiscoveryService.class, immediate = true)
+public class FeicanDiscoveryService extends AbstractDiscoveryService {
+
+    private static final int DISCOVERY_TIMEOUT_SECONDS = 5;
+    private static final int RECEIVE_JOB_TIMEOUT = 20000;
+    private static final int UDP_PACKET_TIMEOUT = RECEIVE_JOB_TIMEOUT - 50;
+    private static final String FEICAN_NAME_PREFIX = "FC_";
+
+    private final Logger logger = LoggerFactory.getLogger(FeicanDiscoveryService.class);
+
+    ///// Network
+    private final byte[] buffer = new byte[32];
+    private final DatagramPacket receivePacket;
+    private final DatagramPacket discoverPacket;
+    @Nullable
+    private DatagramSocket discoverSocket;
+
+    public FeicanDiscoveryService() throws UnknownHostException {
+        super(SUPPORTED_THING_TYPES_UIDS, DISCOVERY_TIMEOUT_SECONDS, false);
+        final InetAddress broadcast = InetAddress.getByName("255.255.255.255");
+        discoverPacket = new DatagramPacket(Commands.discover(), Commands.discover().length, broadcast,
+                Connection.FEICAN_SEND_PORT);
+        receivePacket = new DatagramPacket(buffer, buffer.length);
+    }
+
+    @Override
+    protected void startScan() {
+        logger.debug("Start scan for Feican devices.");
+        discoverThings();
+    }
+
+    @Override
+    protected void stopScan() {
+        logger.debug("Stop scan for Feican devices.");
+        closeDiscoverSocket();
+        super.stopScan();
+    }
+
+    /**
+     * Performs the actual discovery of Feican devices (things).
+     */
+    private void discoverThings() {
+        try {
+            // No need to call close first, because the caller of this method already has done it.
```

Comment:
Can you really be sure of that?

Reply:
It's only called from `startScan` within `AbstractDiscoveryService` which calls `stopScan` before and it's double synchronized (method level and synchronized block) so I assumed I can rely on that.

Reply:
ok!